### PR TITLE
[GOSSIP-63] added CRDT G-Counter implementation

### PR DESCRIPTION
**Pull request by Mirage20**

### CODE REVIEW

File: src/main/java/org/apache/gossip/crdt/CrdtCounter.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gossip.crdt;
+
+import java.util.Set;
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/CrdtCounter.java]: Remove unused imports

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
+package org.apache.gossip.crdt;
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java:19]: Include the apache header in all .java files

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        this.counters.putIfAbsent(other.myID, other.counters.get(other.myID));
+        Map<String , Long> updatedCounter = new HashMap<>();
+        for (Map.Entry<String, Long> entry : this.counters.entrySet()) {
+            String key = entry.getKey();
+            Long value = entry.getValue();
+
+            if(other.counters.containsKey(key)){
+                Long newValue = Math.max(value,other.counters.get(key));
+                updatedCounter.put(key,newValue);
+            }else {
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: Fix formatting you can use the eclipse template (XML) in the project home directory.

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    @Override
+    public GrowOnlyCounter optimize() {
+        return new GrowOnlyCounter(myID, counters.get(myID));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: Avoid random line breaks

File: src/main/java/org/apache/gossip/crdt/CrdtModule.java [hunk 5]
```diff
@@ -42,6 +42,13 @@
   @JsonIgnore abstract boolean isEmpty();
 }
 
+abstract class GrowOnlyCounterMixin{
+  @JsonCreator
+  GrowOnlyCounterMixin(@JsonProperty("myID") String myID, @JsonProperty("counters") Map<String , Long> count){ }
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/CrdtModule.java]: I do not think we should serialize myId. The reason: if one node sets myID and sends that data to another system myId is now incorrect. We need some way to properly hydrate this.

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private GrowOnlyCounter(String myID, Map<String, Long> counters) {
+        this.myID = myID;
+        this.counters.putAll(counters);
+    }
+
+
+    @Override
+    public GrowOnlyCounter merge(GrowOnlyCounter other) {
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: For the other Crdt types they are immutable and merge returns a new instance. I would follow that pattern.

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                updatedCounter.put(key,value);
+            }
+
+        }
+
+        return new GrowOnlyCounter(myID,updatedCounter);
+    }
+
+    @Override
+    public Long value() {
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: 2 space indents

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.lang.annotation.ElementType;
+import java.util.HashMap;
+import java.util.Map;
+
+public class GrowOnlyCounter implements CrdtCounter<Long, GrowOnlyCounter> {
+
+
+    private final Map<String, Long> counters = new HashMap<>();
+    private final String myID;
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: Comment above about myId it should be myId not myID and I think we wish it to be transient or not serialized as it crosses the wire.

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Override
+    public boolean equals(Object obj) {
+
+        GrowOnlyCounter other = (GrowOnlyCounter) obj;
+
+        return value().longValue() == other.value().longValue();
+    }
+
+    public void increase() {
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: Type should be immutable remove this.

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        counters.replace(myID, counters.get(myID) + 1);
+    }
+
+    public void increaseBy(int count) {
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: Type should be immutable remove this.

File: src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        this.myID = myID;
+        counters.putIfAbsent(myID, 0L);
+    }
+
+    private GrowOnlyCounter(String myID, Long count) {
+        this.myID = myID;
+        counters.putIfAbsent(myID, count);
+    }
+
+    private GrowOnlyCounter(String myID, Map<String, Long> counters) {
```
Comment (edwardcapriolo) [inline src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java]: I do not think we want users to manage the myId themselves. I think the interface we should provide is  a builder like OrSet

public GrowOnlyCounter(GrowOnlyCounter g, new GrowOnlyBuilder().setGosipManager(this).increment(1).decrement(3)  ){ }

With reference to GossipManager you can call gossipManager.getMyself().getId() to use as the id

**Review by edwardcapriolo — CHANGES_REQUESTED** [review]

**Comment by Mirage20** [comment]
This pull request was moved to https://github.com/apache/incubator-gossip/pull/42 with updated formatting

**Pull request closed without merging**